src/metabase/driver/mysql/ddl.clj

```diff
@@ -77,68 +77,66 @@
-                                (undofn conn))
-                              (catch Exception _e
-                                (log/warn (trs "Unable to rollback database check for model persistence"))))
-                            step))]
-          (cond (and (= result ::valid) remaining)
-                (recur remaining (conj undo-steps [step undofn]))
+        (let [result (try
+                       (stepfn conn)
+                       (log/info (trs "Step {0} was successful for db {1}" step (:name database)))
+                       ::valid
+                       (catch Exception e
+                         (log/warn (trs "Error in `{0}` while checking for model persistence permissions." step))
+                         (log/warn e)
+                         (try
+                           (doseq [[undo-step undofn] (reverse undo-steps)]
+                             (log/warn (trs "Undoing step `{0}` for db {1}" undo-step (:name database)))
+                             (undofn conn))
+                           (catch Exception _e
+                             (log/warn (trs "Unable to rollback database check for model persistence"))))
+                         step))]
+          (cond (and (= result ::valid) remaining) (recur remaining (conj undo-steps [step undofn]))
+                (= result ::valid) [true :persist.check/valid]
+                :else [false step]))))))
 
-                (= result ::valid)
-                [true :persist.check/valid]
-
-                :else
-                [false step]))))))
+(defmethod ddl.i/check-can-persist :mysql
+  [database]
+  (let [schema-name (ddl.i/schema-name database (public-settings/site-uuid))
+        table-name  (format "persistence_check_%s" (rand-int 10000))
+        steps       [[:persist.check/create-schema
+                      (fn check-schema [conn]
+                        (let [existing-schemas (->> ["select schema_name from information_schema.schemata"]
+                                                    (sql.ddl/jdbc-query conn)
+                                                    (map :schema_name)
+                                                    (into #{}))]
+                          (or (contains? existing-schemas schema-name)
+                              (sql.ddl/execute! conn [(sql.ddl/create-schema-sql database)]))))
+                      (fn undo-check-schema [conn]
+                        (sql.ddl/execute! conn [(sql.ddl/drop-schema-sql database)]))]
+                     [:persist.check/create-table
+                      (fn create-table [conn]
+                        (let [q [(sql.ddl/create-table-sql
+                                  database
+                                  {:table-name        table-name
+                                   :field-definitions [{:field-name "field"
+                                                        :base-type  :type/Text}]}
+                                  "select 1")]]
```

Comment:
Using execute instead of an async call. The async call will produce false results in the following checks.

Reply:
I wonder why we implemented timeouts in the first place. Unless we know this is not needed now, I would prefer fixing async execution such that exceptions are not getting swallowed.

Reply:
@snoe, looks like you were the implementor of the `ddl.i/check-can-persist :mysql` method in question. Any reason why we want to do the create-table check async and the rest sync? I would think for this particular operation, we would need to execute sychronously unless we want to make the entire set of checks async and do a chain of async-await calls, but that seems like major overkill.

Reply:
@markbastian this is not executing concurrently with the other steps, as far as I can tell. The async stuff is there so that we can interrupt it if it takes too long. The problems seems to be that exceptions are not returned from the thread executing the DDL operation.

Reply:
Fixed it to use the older version.